Further centralize shared Fx and TFM transition workarounds

dougbu opened the pull request:
- import generated `Directory.Build.*` files in regular build
  - add `$(UpdateAspNetCoreKnownFramework)` controlling Microsoft.AspNetCore.App KnownFrameworkReference update
  - warn if requested Microsoft.AspNetCore.App version does not exist in `$(DOTNET_ROOT)`
  - remove now-duplicate property and item settings outside generated `Directory.Build.*` files
  - remove central `$(BuildingTestAppsIndependently)` special case
    - not needed because compiler toolset version is always available
  - correct `$(KnownAppHostPackOrFrameworkReferenceTfm)` when not targeting the default TFM
- mention errors with missing generated files and the new warning in BuildFromSource.md
  - address numerous Markdown lint warnings, typos, and spelling mistakes in this file
- use generated `Directory.Build.*` files for local template tests
  - move Infrastructure/ files to TestInfrastructure/
    - Infrastructure/ sub-directories were functionally identical
  - move shared parts of template test project files to PrepareForTest.targets
- use generated `Directory.Build.*` files for local Razor tests
  - set `$(TargetLatestRuntimePatch)` instead of `$(RuntimeFrameworkVersion)`; simpler

dougbu committed:
Further centralize shared Fx and TFM transition workarounds
- make generated `Directory.Build.*` files more widely applicable
  - warn if requested Microsoft.AspNetCore.App version does not exist in `$(DOTNET_ROOT)`
  - add `$(UpdateAspNetCoreKnownFramework)` for Microsoft.AspNetCore.App `@(KnownFrameworkReference)` update
  - remove central `$(BuildingTestAppsIndependently)` special case
    - not needed because compiler toolset version is always available
  - correct `$(KnownAppHostPackOrFrameworkReferenceTfm)` when not targeting the default TFM
    - use MSBuild intrinsic functions for this and in framework projects; future-proofing
  - correct `@(KnownFrameworkReference)` metadata when in servicing
    - should not override default runtime and targeting pack versions
- use generated `Directory.Build.*` files in regular build
  - remove now-duplicate property and item settings outside generated `Directory.Build.*` files
- use generated `Directory.Build.*` files for local Razor tests
  - set `$(TargetLatestRuntimePatch)` instead of `$(RuntimeFrameworkVersion)`; simpler
  - do not restore Razor SDK test asset projects until just before tests run
  - depend on Microsoft.AspNetCore.App projects
  - disable `$(TreatWarningsAsErrors)` for a few Razor SDK tests
    - tests expect projects to build successfully despite a few warnings
- improve (widen) Microsoft.AspNetCore.App `Condition` in Blazor SDK tests

nit: do not pass `$(MicrosoftNetCompilersToolsetVersion)` into Razor test asset projects
  - not needed because generated files already contain the right information
  - even without that, the Directory.Build.props file imports eng/Versions.props

dougbu commented on Directory.Build.targets:
```diff
@@ -144,34 +144,20 @@
+    <!-- Do not update the Microsoft.AspNetCore.App KnownFrameworkReference item unless requested. -->
+    <UpdateAspNetCoreKnownFramework>false</UpdateAspNetCoreKnownFramework>
+    <UpdateAspNetCoreKnownFramework Condition=" '$(UseAspNetCoreSharedRuntime)' == 'true' AND
+      '$(DoNotApplyWorkaroundsToMicrosoftAspNetCoreApp)' != 'true' ">true</UpdateAspNetCoreKnownFramework>
```
This is about the only place the build gets a bit more complicated. Needed to avoid hard-coding "netcoreapp[678...]" somewhere or getting weirdly broken the next time our TFMs change format e.g. if we decide to build for `net8.0-windows`, `net8.0-linux`, and `net8.0-osx` in parallel.

Nice of the SDK to provide the handy-dandy TFM parsing functions though😺

dougbu replied:
Shoot the new `msbuild` functions don't work w/ the old VS on our CI agents ☹️

dougbu commented on docs/BuildFromSource.md:
````diff
@@ -15,13 +15,13 @@ This tutorial assumes that you are familiar with:
 ASP.NET Core uses git submodules to include the source from a few other projects. In order to pull the sources of the these submodules when cloning the repo, be sure to pass the `--recursive` flag to the `git clone` command.
 
-```ps1
+```powershell
````
This file wasn't really incredibly broken; it just gave the VS Code Markdown extension plenty to complain about

dougbu commented on docs/BuildFromSource.md:
```diff
@@ -45,104 +45,109 @@ Visual Studio 2019 (16.8) is required to build the repo locally. If you don't ha
-To set up your project for development on Visual Studio, you'll need to execute the following command.
+To set up your project for development on Visual Studio, you'll need to execute the following command. Building
+subsets of the repo e.g. Java projects may (depending on your scenario) be necessary before building within Visual
+Studio because those projects are not listed in AspNetCore.sln.
```
new

captainsafia replied:
> Studio because those projects are not listed in AspNetCore.sln.

Curious: why is this the case?

BrennanConroy replied:
Nope, no clue how to add javaproj or npmproj to VS.

Nit: "Npm projects" instead of "Java projects" above, no one depends on a Java project.

dougbu commented on docs/BuildFromSource.md:
```diff
@@ -154,26 +159,29 @@ to open the .sln file or one of the project specific .slnf files to work on the
 
-Typically, you want to focus on a single project within the monorepo. For example,
+> :bulb: Rerunning the above command or, perhaps, the quicker `.\build.cmd -noBuildNative -noBuildManaged` may be
+> necessary after switching branches, especially if the `$(DefaultNetCoreTargetFramework)` value changes.
```
new

dougbu commented on docs/BuildFromSource.md:
```diff
@@ -298,23 +304,27 @@ Configuration            | `Debug` or `Release`. Default = `Debug`.
-### Use the result of your build
+### Resx files
+
+If you need to make changes to a .resx file, run `dotnet msbuild t:/Resgen <path to csproj>`. This will update the generated C#.
```
Slight reordering

dougbu commented on docs/BuildFromSource.md:
```diff
@@ -392,3 +398,33 @@ When attempting to restore servicing tags e.g. `v3.1.7`,  the NuGet.config file
 
 The `darc-int-...` feeds in NuGet.config are used only when building internally and are not needed after the tags are created. Delete all such entries in the file and retry.
+
+### Common error: error MSB4019: The imported project "&hellip;\artifacts\bin\GenerateFiles\Directory.Build.props" was not found
```
Wondering whether it makes sense to delete the entire **Troubleshooting** section or reduce it down to a pointer to the BuildErrors.md document. Think I should either duplicate these errors in the two files or do that. Thoughts❔

captainsafia replied:
I think we should move the trouble-shooting section to BuildErrors and keep this doc focused on a build tutorial.

dougbu commented on docs/BuildFromSource.md:
```diff
@@ -392,3 +398,33 @@ When attempting to restore servicing tags e.g. `v3.1.7`,  the NuGet.config file
+### Common error: error MSB4019: The imported project "&hellip;\artifacts\bin\GenerateFiles\Directory.Build.props" was not found
+
+You have likely attempted to use `dotnet` to build a project in the repo without first generating the required files.
+Prior to building without `.\build.cmd` or `.\build.sh` at least the following **must** be executed:
```
```suggestion
Prior to building without `.\build.cmd` or `./build.sh` at least the following **must** be executed:
```

dougbu commented on docs/BuildFromSource.md:
````diff
@@ -392,3 +398,33 @@ When attempting to restore servicing tags e.g. `v3.1.7`,  the NuGet.config file
+or
+
+```bash
+.\build.sh --projects "$PWD/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj"
````
```suggestion
./build.sh --projects "$PWD/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj"
```

dougbu commented on eng/tools/GenerateFiles/Directory.Build.props.in:
```diff
@@ -1,7 +1,6 @@
 <Project>
   <PropertyGroup>
     <DefaultNetCoreTargetFramework>${DefaultNetCoreTargetFramework}</DefaultNetCoreTargetFramework>
-    <TargetFramework>${DefaultNetCoreTargetFramework}</TargetFramework>
```
We don't have any projects lacking a `$(TargetFramework)` setting and this can cause problems where we use `$(TargetFrameworks)`

dougbu commented on eng/tools/GenerateFiles/Directory.Build.targets.in:
```diff
@@ -26,19 +29,27 @@
       LatestRuntimeFrameworkVersion="${MicrosoftNETCoreAppRuntimeVersion}"
```
@ericstj and @Pilchie question that's not directly related to this PR:

We set up this `@(KnownFrameworkReference)` so that we always compile against the latest available targeting pack and shared framework from dotnet/runtime. That's probably fine for release/5.0 unless we service Microsoft.NETCore.App.Ref. But, if we do, I think this approach will set a new minimum shared framework version e.g. 5.0.2 in any package where we ship generated `*.deps.json` or `*.runtimeconfig.json` files. Is that something we should worry about❔

dougbu commented on src/Razor/test/testassets/Directory.Build.props:
```diff
@@ -2,21 +2,17 @@
     <BlazorWebAssemblyJSPath>$(MSBuildThisFileDirectory)blazor.webassembly.js</BlazorWebAssemblyJSPath>
+
+    <!-- Build these test projects against the incoming Microsoft.NETCore.App framework. -->
+    <TargetLatestRuntimePatch>true</TargetLatestRuntimePatch>
```
@dsplaisted and @wli3 given we're setting `%(DefaultRuntimeFrameworkVersion)` to be identical to `%(LatestRuntimeFrameworkVersion)` for the Microsoft.NETCore.App `@(KnownFrameworkReference)` item (in the generated Directory.Build.targets file imported later), does this setting (or making `$(RuntimeFrameworkVersion)` equal to that metadata) change anything at all❔

See https://github.com/dotnet/aspnetcore/pull/27473/files#diff-9fb67cbc0a7ba3b793455ec41e7f36e1d019d5152674d669551810b308510d37R26-R30

wli3 replied:
@dsplaisted double check, i believe it is the same

dsplaisted replied:
No, in this case setting `TargetLatestRuntimePatch` to true will get the version from the "latest" instead of the "default" metadata value, but if they are both the same then it won't affect everything.

dougbu commented:
@captainsafia I'd appreciate your thoughts on docs/BuildFromSource.md changes in particular

captainsafia commented on docs/BuildFromSource.md:
```diff
@@ -154,26 +159,29 @@ to open the .sln file or one of the project specific .slnf files to work on the
 
 Instead, we have many Solution Filter (.slnf) files which include a sub-set of projects. See the Visual Studio
-documentation [here](https://docs.microsoft.com/en-us/visualstudio/ide/filtered-solutions?view=vs-2019) for more
+documentation [here](https://docs.microsoft.com/visualstudio/ide/filtered-solutions?view=vs-2019) for more
```
While we're at it...

```suggestion
documentation [here](https://docs.microsoft.com/visualstudio/ide/filtered-solutions) for more
```

captainsafia reviewed (approved):
BuildFromSource.md changes look good. Left a few notes inline.

dougbu committed:
Apply suggestions from code review

- thanks @captainsafia

Co-authored-by: Safia Abdalla <safia@microsoft.com>

dougbu committed:
Move all troubleshooting information into BuildErrors.md
- some was already duplicated
- fix Markdown lint issues in BuildErrors.md too

dougbu committed:
Switch to .Pre queues for Windows build agents
- required for the new `msbuild` intrinsics
    - it is _not_ recommended to parse TFMs manually anymore
- revert this after Visual Studio 16.8 reaches the regular agents

dougbu committed:
Reorder App.Runtime build slightly
- move `_InstallFrameworkIntoLocalDotNet` earlier because other builds depend on this part
  - this target sometimes executes after dependent projects continue
- add `DependsOnTargets` attributes to further constrain ordering

nit: `IncludeFrameworkListFile` should run before `_ResolveSharedFrameworkContent`

dougbu committed:
Use generated `Directory.Build.*` files for local template tests
- move Infrastructure/ files to TestInfrastructure/
  - Infrastructure/ sub-directories were functionally identical
- move shared parts of template test project files to PrepareForTest.targets

dougbu committed:
Describe errors with missing generated files and the new warning in BuildFromSource.md
- address numerous Markdown lint warnings, typos, and spelling mistakes in this file

dougbu commented on eng/tools/GenerateFiles/Directory.Build.targets.in:
```diff
@@ -2,43 +2,65 @@
+      BeforeTargets="Restore;Build;Rebuild;RunTests;Test;VSTest;Pack"
+      Condition=" $(UpdateAspNetCoreKnownFramework) ">
+    <Warning Text="Requested Microsoft.AspNetCore.App v${MicrosoftAspNetCoreAppRuntimeVersion} does not exist."
+        Condition="!EXISTS('$(DOTNET_ROOT)\shared\Microsoft.AspNetCore.App\${MicrosoftAspNetCoreAppRuntimeVersion}')" />
```
@rainersigwald we hit this warning in the `RestoreTestProjects` targets in src/Components/WebAssembly/Sdk/integrationtests/Microsoft.NET.Sdk.BlazorWebAssembly.IntegrationTests.csproj and src/Razor/Microsoft.NET.Sdk.Razor/integrationtests/Microsoft.NET.Sdk.Razor.IntegrationTests.csproj though those targets execute a while after the src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj has laid out the newly-built shared framework in this directory. Is the `EXISTS` check using a cache of the directory content from earlier in the build❔ If yes, what's the workaround (I prefer to avoid separating those test projects into another build step)❔ If no, what am I doing wrong❔

rainersigwald replied:
Exists as a condition inside a target should happen late and not be cached. I'm not aware of any bugs in that area.

How are you determining the order of operations here? Is there a binlog uploaded in the build?

dougbu replied:
I reproduced the problem in a local build and used the timeline view in that binary log. Because this requires a clean repo, including removal of 6.0.0-dev folders under .dotnet/, the binary log is pretty large❕ But, I'll do it again and share the results this afternoon&hellip;

dougbu replied:
Bit more detail:
``` text
> dotnet msbuild .\artifacts\log\Debug\Build.binlog /noconlog '/flp:v=diag;logfile=diag.log'
...
> sls '-\>.*\\.dotnet\\shared', 'error :' .\diag.log
diag.log:39258399:                     Task Parameter:Text=Microsoft.AspNetCore.App.Runtime.csproj -> ...\.dotnet\shared\Microsoft.AspNetCore.App\6.0.0-dev\ (TaskId:7247)
diag.log:39258400:                     Microsoft.AspNetCore.App.Runtime.csproj -> ...\.dotnet\shared\Microsoft.AspNetCore.App\6.0.0-dev\ (TaskId:7247)
diag.log:57244108:19:33:52.124   529>...\artifacts\bin\GenerateFiles\Directory.Build.targets(74,5): error : Requested Microsoft.AspNetCore.App v6.0.0-dev does not exist. [...\src\Components\WebAssembly\Sdk\testassets\RestoreBlazorWasmTestProjects\RestoreBlazorWasmTestProjects.csproj]
diag.log:57244113:19:33:52.124   529>...\artifacts\bin\GenerateFiles\Directory.Build.targets(74,5): error : Requested Microsoft.AspNetCore.App v6.0.0-dev does not exist. [...\src\Components\WebAssembly\Sdk\testassets\RestoreBlazorWasmTestProjects\RestoreBlazorWasmTestProjects.csproj]
diag.log:65080699:19:37:05.409   535>...\artifacts\bin\GenerateFiles\Directory.Build.targets(74,5): error : Requested Microsoft.AspNetCore.App v6.0.0-dev does not exist.
...
> . .\activate.ps1
Enabled the .NET Core environment. Execute 'deactivate' to exit.
dotnet = C:\dd\dnx\aspnetcore\.dotnet\dotnet.exe
> dir "$env:DOTNET_ROOT\shared\Microsoft.AspNetCore.App\6.0.0-dev\"
    Directory: ...\.dotnet\shared\Microsoft.AspNetCore.App\6.0.0-dev

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a---          2020-11-17    19:30             53 .version
-a---          2020-11-17    19:13        1282048 aspnetcorev2_inprocess.dll
-a---          2020-11-17    19:25          48640 Microsoft.AspNetCore.Antiforgery.dll
-a---          2020-11-17    19:30          21393 Microsoft.AspNetCore.App.deps.json
...
```

dougbu replied:
The 6.0.0-dev directory was last updated at 19:30:38, a few minutes before the problems started at 19:33:52.

dougbu replied:
@rainersigwald ignore all of the above. The issue was my incorrect assumption `$(DOTNET_ROOT)` was always set. Noticing the problem went away in _some_ of my local builds and some time up close and personal with binary logs helped (eventually).

dougbu committed:
Add temporary workaround for `[AssemblyVersion]` changes

dougbu commented:
@wtgodbe and @JunTaoLuo please review. Lots has changed since @captainsafia signed off.

dougbu commented on eng/tools/GenerateFiles/Directory.Build.targets.in:
```diff
@@ -2,43 +2,92 @@
+      <DefaultRuntimeFrameworkVersion Condition=" '$(IsServicingBuild)' != 'true' AND
+          '$(TargetLatestDotNetRuntime)' != 'false' ">${MicrosoftNETCoreAppRuntimeVersion}</DefaultRuntimeFrameworkVersion>
+      <TargetingPackVersion Condition=" '$(IsServicingBuild)' != 'true' AND
+          '$(TargetLatestDotNetRuntime)' != 'false' ">${MicrosoftNETCoreAppRefVersion}</TargetingPackVersion>
```
@jkotalik the `AND '$(TargetLatestDotNetRuntime)' != 'false'` additions here are what #27845 needs (in root Directory.Build.targets) until this gets in.

jkotalik replied:
Do I need to these changes specifically in this file as well? Sounds like no.

dougbu committed:
Address @wtgodbe's nit from #27653
- https://github.com/dotnet/aspnetcore/pull/27653#pullrequestreview-529368502

wtgodbe reviewed (commented):
Assuming this works, it seems reasonable, but I'm concerned that it seems to be adding complexity and might make our build even harder to understand than it already is - could you explain if/how the value it adds is large enough to overcome that (or if I'm overestimating the added complexity)?

dougbu commented:
> could you explain if/how the value it adds is large enough to overcome that (or if I'm overestimating the added complexity)?

I started this branch to

1. remove duplication of very similar workarounds sprinkled around the repo
  - this should make it easier to correct problems in an area we've had some churn and late-breaking concerns
2. avoid problems such as #14895

The work also helped me remove duplication I was less aware of e.g. between src/ProjectTemplates/BlazorTemplates.Tests/ and src/ProjectTemplates/test/.

What specific added complexity are you concerned about❔ I suspect at least some if what you're seeing existed before but in code I'm deleting.

dougbu commented:
@wtgodbe this is now green. Do you object to merging❔

wtgodbe commented:
> What specific added complexity are you concerned about❔ I suspect at least some if what you're seeing existed before but in code I'm deleting.

Main concern is the use of the generated .props/.targets file everywhere now, it seems a bit unintuitive and may leave devs wondering where certain common properties "live" now

dougbu commented:
>> What specific added complexity are you concerned about❔ I suspect at least some if what you're seeing existed before but in code I'm deleting.

> Main concern is the use of the generated .props/.targets file everywhere now, it seems a bit unintuitive and may leave devs wondering where certain common properties "live" now

For me, that intuition is offset by the ability to easily search the repo and the reduced distribution of settings. We've had too many cases in the past where "central" settings aren't applied in isolated parts of the repo (test asset projects in particular) and someone ends up changing just one location.

But, if you're not ready to sign off, let's discuss in January.

wtgodbe commented:
If I clean the repo then run `dotnet build` on an arbitrary project, will it work/create the generated file correctly pre-build? If so I think I'm ok with this.

wtgodbe commented:
Seems like that works, so I'm satisfied

dougbu commented:
> If I clean the repo then run `dotnet build` on an arbitrary project, will it work/create the generated file correctly pre-build? If so I think I'm ok with this.

`dotnet build` after `git clean -e .dotnet/` won't work for many projects in the repo today. In fact, it'll fail for _every_ project if the .dotnet/ folder is not up-to-date. That's why docs/BuildFromSource.md recommends running ./restore.cmd before doing much else.

If the SDK is up-to-date, the new minimum requirement after cleaning for `dotnet build` (in projects that don't require native assets) is `./build.cmd -noBuildNative -noBuildManaged` or `./build.sh --no-build-managed`. That just builds the repo tasks and generated the eng/Directory.Build.* files and takes a few seconds. I documented that in this PR.

dougbu commented:
> Seems like that works, so I'm satisfied

As I said, that isn't expected to work. I should have added that `./build.* -projects ...` will _always_ work because the command builds native assets (if on Windows), updates the SDK, ensures the needed runtimes are available, as well as generates the artifacts/GenerateFiles/* files.

dougbu commented:
Do those caveats undo your approval❔

wtgodbe commented:
> If the SDK is up-to-date, the new minimum requirement after cleaning for dotnet build (in projects that don't require native assets) is ./build.cmd -noBuildNative -noBuildManaged or ./build.sh --no-build-managed

I did my normal workflow, which is clean, restore, build a project using the restored .dotnet folder. Given that this breaks folks who are running `dotnet build` with a globally installed SDK that matches the one in global.json (e.g. once we ship 6.0 RTM, or if anyone has a 6.0 preview installed globally), I'm more hesitant. My gut feeling is that that scenario should work. Maybe if every `restore` operation creates the generated file if it doesn't already exist? In any case I'd prefer to talk about this more before making that breaking change.

wtgodbe reviewed (changes requested):
> If the SDK is up-to-date, the new minimum requirement after cleaning for dotnet build (in projects that don't require native assets) is ./build.cmd -noBuildNative -noBuildManaged or ./build.sh --no-build-managed

I did my normal workflow, which is clean, restore, build a project using the restored .dotnet folder. Given that this breaks folks who are running `dotnet build` with a globally installed SDK that matches the one in global.json (e.g. once we ship 6.0 RTM, or if anyone has a 6.0 preview installed globally), I'm more hesitant. My gut feeling is that that scenario should work. Maybe if every `restore` operation creates the generated file if it doesn't already exist? In any case I'd prefer to talk about this more before making that breaking change.

dougbu commented:
> I did my normal workflow, which is clean, restore, build a project&hellip;

Ah, I thought you wanted to avoid `./restore.*`. If you `./restore.*` after cleaning, everything is just fine for any project that doesn't require native assets.

> Given that this breaks folks who are running dotnet build with a globally installed SDK

This PR doesn't break using a globally-installed SDK. Globally-installed SDKs only work if they _also_ include up-to-date runtimes from dotnet/runtime e.g. just after rebranding post-release. This restriction comes from the need to have the additional runtimes in the SDK folder. Nothing new here w.r.t. that restriction.

Did you request changes because you thought something about globally-installed SDKs had changed in this PR❔

wtgodbe commented:
So what causes the new requirement to run `./build.cmd -noBuildNative -noBuildManaged or ./build.sh --no-build-managed` before building a single project?

dougbu commented:
> So what causes the new requirement to run `./build.cmd -noBuildNative -noBuildManaged` or `./build.sh --no-build-managed` before building a single project?

I probably I confused you because I thought you wanted to avoid using `./restore.*` and responded with the absolute minimal requirement to build a managed project. Those commands do _less_ than `./restore.*`.

wtgodbe commented:
> I probably I confused you because I thought you wanted to avoid using ./restore.* and responded with the absolute minimal requirement to build a managed project. Those commands do less than ./restore.*.

I understand that part, I just want to understand what (if any) scenarios worked before this PR, that do not work after - that is, who has to change their workflow?

dougbu commented:
> what (if any) scenarios worked before this PR, that do not work after

About the only things that change at all are scenarios involving switches between branches for which all needed runtimes are already in .dotnet/. In those cases, it's better to use `./build.* -projects {path}` instead of or `./build.* -noBuildManaged -noBuildNative` before `dotnet build`. But, the worst case would be using a different Microsoft.NETCore.App and that can only occur if the two branches began at different 'master' commits with dependency updates between.

wtgodbe reviewed (approved):
Ok, then this works for me

dougbu committed:
Revert "Switch to .Pre queues for Windows build agents"
- let's see if 16.8.0 is new enough
- regular queues were updated today

This reverts commit e4ed1755ee74565b232bf0d08b7c631a6e775778.
